### PR TITLE
Extract code from Optim.

### PR DESCRIPTION
**Pull request by pkofod**
The question is; how much of the actual functionality should be here? Enough to allow sharing of api and types (I guess?) between Optim and LineSearches (and NLsolve), but is it "too much" to have all the constructors (for autodiff) here as well? This is very much up for discussion.

### CODE REVIEW

**Comment by pkofod** [comment]
@KristofferC @anriseth

**Comment by anriseth** [comment]
From LineSearches' point of view, the autodiff constructers would not be needed.
The gradient testing suite may find the finitediff stuff useful though (https://github.com/JuliaNLSolvers/Optim.jl/issues/386)

File: src/objective_types.jl [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+
+function _unchecked_grad!(obj, x)
+    obj.g_calls .+= 1
+    copy!(obj.last_x_g, x)
+    obj.g!(x)
+end
+function gradient!(obj::AbstractObjective, x)
+    if x != obj.last_x_g
+        _unchecked_gradient!(obj, x)
```
Comment (anriseth) [inline src/objective_types.jl]: Should be `_unchecked_grad!`?

Comment (pkofod) [inline src/objective_types.jl]: sure. I guess this never gets called  :)

**Comment by pkofod** [comment]
I also want to discuss: do we want `grad' or should we just write it out as `gradient`? What about for NLsolve, there it should be Jacobian, so I guess we should pull in the relevant code from that package as well?

File: src/objective_types.jl [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+        _unchecked_value!(obj, x)
+    end
+    obj.f_x
+end
+
+
+function _unchecked_grad!(obj, x)
+    obj.g_calls .+= 1
+    copy!(obj.last_x_g, x)
+    obj.g!(x)
```
Comment (anriseth) [inline src/objective_types.jl]: And this should be `obj.g!(x, obj.g)` I guess? [Or the parameters the other way around when we change the order]

Comment (pkofod) [inline src/objective_types.jl]: Yes, you're right, thanks. I actually did notice this myself, but apparently forgot to change it. They aren't used (yet) as all gradient updates happen through `fg!`.

File: src/objective_types.jl [hunk 3]
```diff
@@ -228,18 +228,18 @@ function value!(obj, x)
 end
 
 
-function _unchecked_grad!(obj, x)
+function _unchecked_gradient!(obj, x)
     obj.g_calls .+= 1
     copy!(obj.last_x_g, x)
-    obj.g!(x)
+    obj.g!(x, stor)
```
Comment (anriseth) [inline src/objective_types.jl]: There is no `stor` argument here, do you mean `obj.g`? Alternatively we'd have to add a `storage` parameter to `gradient!` and `value_gradient!` etc. as well.

Comment (pkofod) [inline src/objective_types.jl]: ! 

...

(don't code sleep deprived)

**Pull request merged**